sdk/helper/aliasmetadata/alias_metadata.go

```diff
@@ -0,0 +1,212 @@
+	// include the "default" parameter, and instead includes the actual
+	// fields behind "default", if selected. If it has never been set
+	// or only the default fields are desired, it is nil.
+	aliasMetadata []string
+
+	// fields is a list of the configured default and available
+	// fields. It's intentionally not jsonified.
+	fields *Fields
+}
+
+// AliasMetadata is intended to be used on config reads.
+// It gets an explicit list of all the user's configured
+// fields that are being added to alias metadata. It will never
+// include the "default" parameter, and instead includes the actual
+// fields behind "default", if selected.
+func (h *Handler) AliasMetadata() []string {
+	if h.aliasMetadata == nil {
+		return h.fields.Default
+	}
+	return h.aliasMetadata
+}
+
+// ParseAliasMetadata is intended to be used on config create/update.
+// It takes a user's selected fields (or lack thereof),
+// converts it to a list of explicit fields, and adds it to the Handler
+// for later storage.
+func (h *Handler) ParseAliasMetadata(data *framework.FieldData) error {
+	userProvidedRaw, ok := data.GetOk(h.fields.FieldName)
+	if !ok {
+		// Nothing further to do here.
+		return nil
+	}
+	userProvided, ok := userProvidedRaw.([]string)
+	if !ok {
+		return fmt.Errorf("%s is an unexpected type of %T", userProvidedRaw, userProvidedRaw)
+	}
+
+	// If the only field the user has chosen was the default field,
+	// we don't store anything so we won't have to do a storage
+	// migration if the default changes.
+	if len(userProvided) == 1 && userProvided[0] == "default" {
+		h.aliasMetadata = nil
+		return nil
+	}
+
+	userProvided = h.expandDefaultField(userProvided)
+	userProvided = strutil.RemoveDuplicates(userProvided, true)
+	if !strutil.StrListSubset(h.fields.all(), userProvided) {
+		return fmt.Errorf("%q contains an unavailable field, please select from %q",
+			strings.Join(userProvided, ", "), strings.Join(h.fields.all(), ", "))
+	}
+	h.aliasMetadata = userProvided
+	return nil
+}
+
+// expandDefaultField looks for the field "default" and, if exists,
+// replaces it with the actual default fields signified.
+func (h *Handler) expandDefaultField(userProvided []string) (expanded []string) {
+	for _, field := range userProvided {
+		if field != "default" {
+			expanded = append(expanded, field)
+			continue
+		}
+		for _, dfltField := range h.fields.Default {
+			expanded = append(expanded, dfltField)
+		}
+	}
+	return expanded
+}
+
+// PopulateDesiredAliasMetadata is intended to be used during login
+// just before returning an auth.
+// It takes the available alias metadata and,
+// if the auth should have it, adds it to the auth's alias metadata.
+func (h *Handler) PopulateDesiredAliasMetadata(auth *logical.Auth, available map[string]string) error {
+	if auth == nil {
+		return errors.New("auth is nil")
+	}
+	if auth.Alias == nil {
+		return errors.New("auth alias is nil")
+	}
+	if auth.Alias.Name == "" {
```

Comment:
Is `Alias.Name` is related to the metadata handling? This seems to be a concern for the framework, which is processing the Alias.

Reply:
Sure, I could leave that out, I suppose it could be useful in case they're not adding everything in order.

Reply:
This is addressed in https://github.com/hashicorp/vault/pull/8783/commits/ae3ba7f04ddcac3eb4f03036e4a6c2ce4028081e